lib/state/selectors.ts

```diff
@@ -58,3 +59,43 @@ export const isDialogOpen = (state: S.State, name: T.DialogType['type']) =>
 
 export const numberOfNonEmailTags: S.Selector<number> = ({ data }) =>
   [...data.tags.values()].filter((tag) => !isEmailTag(tag.name)).length;
+
+export const noteTags: S.Selector<T.Tag[]> = ({ data }, note: T.Note) => {
+  return note.tags
+    .filter((tagName) => !isEmailTag(tagName))
+    .map(
+      (tagName) =>
+        data.tags.get(tagHashOf(tagName)) ?? {
+          name: tagName,
+          deleted: true,
+        }
+    );
+};
+
+export const getRevision: S.Selector<T.Note | null> = (
+  state,
+  noteId: T.EntityId,
+  revisionVersion: number,
+  includeDeletedTags: boolean
+) => {
+  const note = state.data.notes.get(noteId);
+  const revisions = state.data.noteRevisions.get(noteId);
+  const revision = revisions?.get(revisionVersion);
+
+  if (!note || !revision) {
+    return null;
+  }
+
+  const noteEmailTags = note.tags.filter((tagName) => isEmailTag(tagName));
+  const revisionCanonicalTags = revision.tags.filter((tagName) => {
+    const tagHash = tagHashOf(tagName);
+    const hasTag = state.data.tags.has(tagHash);
+    return !isEmailTag(tagName) && (hasTag || includeDeletedTags);
+  });
+
+  return {
+    ...revision,
+    tags: [...noteEmailTags, ...revisionCanonicalTags],
+    systemTags: note.systemTags,
+  };
```

Comment:
this is a nice approach. I tried thinking through ways we could do without this, passing the `Note` directly, but couldn't find any way to create this revision `Note` without knowing about the current app state. in the end I think your idea here is the best way to resolve our needs. it's a more-complicated task (communicating through the Redux store what needs to be restored) than it seems at face value.